include/sys/onoff.h

```diff
@@ -402,65 +335,109 @@ static inline void onoff_client_init_callback(struct onoff_client *cli,
+ * Cancelling is supported only for onoff_request() and onoff_reset()
+ * operations, and is a synchronous operation.  Be aware that any
+ * transition that was initiated on behalf of the client will continue
+ * to progress to completion: it is only notification of transition
+ * completion that may be eliminated.  If there are no active requests
+ * when a transition to on completes the manager will initiate a
+ * transition to off.
  *
- * @return true if and only if the service has an uncleared error.
+ * Client notification does not occur for cancelled operations.
+ *
+ * @param mgr the manager for which an operation is to be cancelled.
+ *
+ * @param cli a pointer to the same client state that was provided
+ * when the operation to be cancelled was issued.
+ *
+ * @retval non-negative the observed state of the machine at the time
+ * of the cancellation, if the cancellation succeeds.  On successful
+ * cancellation ownership of @c *cli reverts to the client.
+ * @retval -EINVAL if the parameters are invalid.
+ * @retval -EALREADY if cli was not a record of an uncompleted
+ * notification at the time the cancellation was processed.  This
+ * likely indicates that the operation and client notification had
+ * already completed.
  */
-static inline bool onoff_has_error(const struct onoff_manager *mgr)
+int onoff_cancel(struct onoff_manager *mgr,
+		 struct onoff_client *cli);
+
+/**
+ * @brief Helper function to safely cancel a request.
+ *
+ * Some applications may want to issue requests on an asynchronous
+ * event (such as connection to a USB bus) and to release on a paired
+ * event (such as loss of connection to a USB bus).  Applications
+ * cannot precisely determine that an in-progress request is still
+ * pending without using onoff_monitor and carefully avoiding race
+ * conditions.
+ *
+ * This function is a helper that attempts to cancel the operation and
+ * issues a release if cancellation fails because the request was
+ * completed.  This synchronously ensures that ownership of the client
+ * data reverts to the client so is available for a future request.
+ *
+ * @param mgr the manager for which an operation is to be cancelled.
+ *
+ * @param cli a pointer to the same client state that was provided
+ * when onoff_request() was invoked.  Behavior is undefined if this is
+ * a pointer to client data associated with an onoff_reset() request.
+ *
+ * @retval ONOFF_STATE_TO_ON if the cancellation occurred before the
+ * transition completed.
+ *
+ * @retval ONOFF_STATE_ON if the cancellation occurred after the
+ * transition completed.
+ *
+ * @retval -EINVAL if the parameters are invalid.
+ *
+ * @retval negative other errors produced by onoff_release().
+ */
+static inline int onoff_cancel_or_release(struct onoff_manager *mgr,
+					  struct onoff_client *cli)
 {
-	return (mgr->flags & ONOFF_HAS_ERROR) != 0;
+	int rv = onoff_cancel(mgr, cli);
+
+	if (rv == -EALREADY) {
+		rv = onoff_release(mgr);
```

Comment:
client state shall be cleaned here. I see that we are missing a call in `sys_notify` for that. Something like `sys_notify_reset`.

Reply:
See comment above: there is no notion of cleaning client state in this infrastructure.  The client is responsible for initializing it before hand-off to the manager, and for interpreting whatever the manager set when it was returned.  The only thing the manager does is set completion state, if the operation completes before the client data is returned.